### ISSUE TITLE
Not properly handling fs events for uris that are currently in a pending `stat` or `readFile` call

### ISSUE THREAD

**Issue by eamodio** [microsoft/vscode#132765]
Refs: https://github.com/microsoft/vscode-remotehub/issues/274

In the GitHub Repositories extension, I'm seeing an issue where if a `FileChangeType.Changed` `onDidChangeFile` event is trigger for a Uri while a `stat` and/or `readFile` request for that same Uri is in flight, the event seems to be ignored.

My guess is that our fs handling code, is probably handling the event, but because there is already a `stat` call (or `readFile` call) running it just uses the running call, rather than discarding the current call and issuing a new one. While this case is probably extremely rare with normal file systems, because a) most calls are very fast normally, and the chances of an event when we are waiting for them to complete is likely very small. But for GitHub Repos this can happened very frequently (at startup).

During startup in GitHub Repos, we try to figure the commit sha the repository should be set to -- which depends on if you've opened that repository before, and if there are any uncommitted changes (there are other factors as well, but they don't really matter for this discussion). So if you've open this repo before, we will start the repository at the commit you were at the last time you opened the repo, but then if there are no uncommitted changes, we fast-forward you to the tip. And that's where this issue comes into play -- because when we fast-forward the repository, we fire change events for the repository and "watched" uris to ensure any opened file will get refreshed with the latest content. But because the GitHub Repos fs calls can sometimes take a long time to complete, especially during startup (as there is some blocking setup that must complete first), there can be `stat` and/or `readFile` calls in flight when we trigger those events.

In GitHub Repositories, I've kind of "worked" around this issue currently, by keeping track of all pending `stat`/`readFile` requests, and if the repo changes while they are still running, I explicitly trigger another event on a `setTimeout` (of 10ms -- 0 didn't work either for some reason) to that uri when the calls finish. But that code adds quite a bit of unwanted complexity and shouldn't be required.

**Comment by bpasero** [comment]
No, not to my knowledge. The only `queue` the file service maintains is for writing because node.js does not recommend to have multiple `async` writes to the same file at the same time:

https://github.com/microsoft/vscode/blob/3252199273bbd0e179aeee36750a8d6424b77066/src/vs/platform/files/common/fileService.ts#L1167-L1167

I suspect this is not blocked at the file service layer but maybe in the text file model layer that sits on top. We have code that will resolve a opened text model again from file changes that impact it and we queue this up to have at maximum 2 operations: 1 that is running and 1 that is pending:

https://github.com/microsoft/vscode/blob/84dad213a044a74d9481ed6ee35439b06cc1927a/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts#L148-L163

To move forward I need you to build a minimal repro with a hello world extension and some steps how to try this out.

**Comment by bpasero** [comment]
Any update on my question?

**Comment by eamodio** [comment]
@bpasero I've put a super simple fs provider here: https://github.com/eamodio/vscode-simple/tree/fs-events which basically does what RemoteHub does -- mtime & contents are based on a global "revision" (the current sha of the repo)

The built vsix is here: [simple-0.0.1.vsix.zip](https://github.com/microsoft/vscode/files/7282969/simple-0.0.1.vsix.zip)

You can reproduce the issue, by installing the vsix, open a new vscode window, and then run the _Simple: Open Virtual File_ command. This will open `foo.md` from the `simple` fs provider. The contents of the file should be `1` at first and then get updated to `2`. But if you just run the command and do nothing, it will just show `1`, if you unfocus the window and re-focus it (vscode will issue a `stat` call) and that will cause the contents to be fixed, to show `2`.

You can close the `foo.md` file, and run the _Simple: Reset_ command to try again.

**Comment by bpasero** [comment]
This is a combination of a few things: the file that opens is considered to be "out of workspace" at least in your example because no workspace is opened that contains this file. We only watch for changes to "out of workspace" files when we detect an editor opening here:

https://github.com/microsoft/vscode/blob/23f8788a8b167ef8ec597967374c96bd27d456aa/src/vs/workbench/services/editor/browser/editorService.ts#L184

But this only happens when the editor was opened, so it will happen after the file was changed by the extension.

Assuming that in the GH case, this is not a problem because in that case the file is very likely considered to be inside the workspace so we should be getting the file event, there is another place where we update models based on file events but we skip any model that is not resolved:

https://github.com/microsoft/vscode/blob/cb48ce5ca7b083918904c3bf7c489ff9f7c3dc1f/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts#L111

In this case, since the model is still loading, the file event is discarded. I think that is the underlying issue here and I have a branch up that changes this piece of code to allow a model reload even when the model is pending to be resolved. With the idea being to schedule a resolve after the pending one has finished.

@eamodio can you please verify from https://github.com/microsoft/vscode/tree/ben/132765 that the issue is resolved? We cannot really use your test extension as a sample here for the reasons stated above, but I think you should see the fix working for the GH remote case at least.
